### PR TITLE
Add more code-scanning suites

### PR DESCRIPTION
**Pull request by robertbrignull**
We want to add more code scanning query suites so that users to easily choose to run a larger set of query in a fully supported and maintainable way. We also don't want people to have to use the lgtm suites because it's an old product name that people shouldn't have to know about.

Suite names and contents were decided upon by @jhutchings1 and @sj.

Note that the `security-and-quality` suite is effectively the same as the `lgtm-full` suite, but it is missing lgtm-specific things like the `definitions.ql` query. The `security-extended` suite is the again same as this, but limited to only security queries.

For cpp there were a few queries explicitly removed, and I've moved the definition of this list to a separate file so those queries can easily be excluded from the other code scanning suites.

### CODE REVIEW

**Comment by robertbrignull** [comment]
I'm including these suites in a new codeql bundle I'm uploading to code scanning. I'll confirm once this is done and that they're working there.

File: cpp/ql/src/codeql-suites/slow-queries.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+- description: C/C++ queries are infeasible to compute on large projects
```
Comment (rdmarsh2) [inline cpp/ql/src/codeql-suites/slow-queries.yml:1]:
```suggestion
- description: C/C++ queries which are infeasible to compute on large projects
```

**Comment by robertbrignull** [comment]
These suites have been included in https://github.com/github/codeql-action/releases/tag/codeql-bundle-20200601 and they are working as expected.

**Comment by robertbrignull** [comment]
Are the python errors `semmlecode-python-tests/extractor-tests/long_path/LongPath.ql: Test failure (Expected output does not match)` something to worry about? Are they caused by this PR or an existing issue?

**Comment by tausbn** [comment]
> Are the python errors `semmlecode-python-tests/extractor-tests/long_path/LongPath.ql: Test failure (Expected output does not match)` something to worry about? Are they caused by this PR or an existing issue?

They were caused by an existing issue that has since been resolved: https://git.semmle.com/Semmle/code/pull/37071

**Comment by robertbrignull** [comment]
The Go version of this PR (https://github.com/github/codeql-go/pull/163) has been merged. Would be nice to get this one in soon so that repo doesn't have a reference to a file that doesn't exist.

**Comment by jhutchings1** [comment]
@p0 Can you help get this one across the line? 🙇🏻

**Comment by jbj** [comment]
I don't know what I'm being asked to merge or approve here. It's a small diff on top of a suite description DSL that I've never seen. It picks certain queries for certain suites, but I can't say whether those queries are appropriate for those suites since that'll depend on how they are presented to the user.

If we present an extended suite as "here are some more queries, some of which might be interesting to you depending on the nature of your project and depending on what you're looking to find", then I'm sure it's fine to include medium-precision queries. But for C/C++ we have many medium-precision queries that are useless for 90% of all projects.

**Comment by robertbrignull** [comment]
> If we present an extended suite as "here are some more queries, some of which might be interesting to you depending on the nature of your project and depending on what you're looking to find"

I think that's an accurate description. These suites are party for people who want to run more queries, and they're also to satisty demands met by the field team from people who want to run the same queries as on LGTM.

If people find that some of the queries are too slow or not useful they can remove them by writing a suite of their own. That's a lot more effort though so we want to offer a very easy way to run a generic large suites of queries.

> But for C/C++ we have many medium-precision queries that are useless for 90% of all projects.

The largest suite being introduced here is `security-and-quality` which is essentially the same (actually a strict subset) of the `lgtm-full` set which was run by lgtm. I have been assuming that would be an ok group of queries to run as we were already running them on thousands of projects. For cpp you'll note there are already a few queries that are explicitly removed, so the list is manually curated, or has the potential to be so.

If you'd like to suggest a diferent list of queries that would be extremely helpful. I don't want to take power away from the language teams.

File: cpp/ql/src/codeql-suites/slow-queries.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+- description: C/C++ queries which are infeasible to compute on large projects
```
Comment (aibaars) [inline cpp/ql/src/codeql-suites/slow-queries.yml:1]: I think `slow-queries.yml` is a misleading name. This suite excludes slow queries, not include them. I think `exclude-slow-queries.yml` would be clearer.

Comment (aibaars) [reply cpp/ql/src/codeql-suites/slow-queries.yml:1]: @jbj Could these queries be optimised, so they don't need to be excluded? The majority seem related to unclosed resources.

**Comment by aibaars** [comment]
> I don't know what I'm being asked to merge or approve here. It's a small diff on top of a suite description DSL that I've never seen. It picks certain queries for certain suites, but I can't say whether those queries are appropriate for those suites since that'll depend on how they are presented to the user.

@jbj The suite description DSL is part of CodeQL. You can run `codeql resolve queries SUITE_NAME` to see the list of queries the suite would evaluate to.

File: cpp/ql/src/codeql-suites/slow-queries.yml (hunk 2 above)
Comment (p0) [reply cpp/ql/src/codeql-suites/slow-queries.yml:1]: The reason they're not scalable is that they use the old Steensgaard-style pointer analysis. Optimising them would mean rewriting with modern techniques -- which is a worthy goal, of course, but not an easy one without affecting results.

**Comment by robertbrignull** [comment]
Unfortunately dismissed the reviews, but @aibaars and @p0's comments are now addressed.

**Comment by jbj** [comment]
To be clear, I'm not against this PR being merged. If you feel confident merging it, @aibaars, then please do.

I've re-triggered the tests as they were failing due to a CI race condition.

**Comment by aibaars** [comment]
@jbj It looks fine from a Java point of view. The suites are essentially the same as what LGTM is running.

As far as I can tell the only controversial bit is the exclusion of some slow resource management queries. Perhaps you could raise an issue to improve those queries when time permits and get rid of the exclusion list.

**Comment by jbj** [comment]
Okay, this PR has been sitting around for long enough.

**Pull request merged**